### PR TITLE
[ticket/14323] Added support for truncating long URLs

### PR DESCRIPTION
**Pull request by JoshyPHP**
PHPBB3-14323

https://area51.phpbb.com/phpBB/viewtopic.php?f=126&t=48751

### CODE REVIEW

**Comment by JoshyPHP** [comment]
!set WIP

File: phpBB/phpbb/textformatter/s9e/factory.php [hunk 1]
```diff
@@ -395,6 +394,31 @@ public function regenerate()
...
+	* @return void
+	*/
+	protected function configure_autolink(Configurator $configurator)
+	{
+		$configurator->plugins->load('Autoemail');
+		$configurator->plugins->load('Autolink', array('matchWww' => true));
+
+		// Create a tag that will be used to display the truncated text by replacing the original
+		// content with the content of the @text attribute
+		$tag = $configurator->tags->add('TRUNCATED_LINK');
```
Comment (JoshyPHP) [inline phpBB/phpbb/textformatter/s9e/factory.php]: Probably not the best name for it. Maybe something like `AUTOLINK_TEXT` would make more sense.

**Comment by JoshyPHP** [comment]
Truncating local URLs requires to figure out the board's URL. Like the original code `make_clickable()` this patch uses `generate_board_url()`. It adds a lot of complexity in the background because of its dependency on the global request/user. If there's an alternative, it should be replaced. Otherwise, I recommend carefully weighing the value of this functionality (truncating local URLs) against its cost in code complexity.

File: phpBB/phpbb/textformatter/s9e/parser.php [hunk 2]
```diff
@@ -393,4 +393,49 @@ static public function filter_img_url($url, array $url_config, Logger $logger, $
...
+	/**
+	* Replace the content displayed inside of a URL tag
+	*
+	* Will only apply to URL tags that do not use any markup (e.g. not "[url]") on the assumption
+	* that those tags were created by the Autolink plugin to linkify URLs found in plain text
+	*
+	* @param  \s9e\TextFormatter\Parser\Tag $url_tag   URL tag (start tag)
+	* @param  \s9e\TextFormatter\Parser     $parser    Parser
+	* @param  string                        $board_url Forum's root URL (with trailing slash)
+	* @return bool                                     Always TRUE to indicate that the tag is valid
```
Comment (marc1706) [inline phpBB/phpbb/textformatter/s9e/parser.php]: Please just write ```true```. We don't use ```TRUE``` in phpBB.

File: phpBB/phpbb/textformatter/s9e/parser.php [hunk 3]
```diff
@@ -393,4 +393,49 @@ static public function filter_img_url($url, array $url_config, Logger $logger, $
...
+	*/
+	public static function generate_autolink_text(\s9e\TextFormatter\Parser\Tag $url_tag, \s9e\TextFormatter\Parser $parser, $board_url)
+	{
+		// If the tag consumes any text then we ignore it because it's not a linkified URL. Same if
```
Comment (marc1706) [inline phpBB/phpbb/textformatter/s9e/parser.php]: I'd prefer if you tried to use a maximum of 80 chars per line in comments.

File: tests/text_processing/tickets_data/PHPBB3-10587.html [hunk 4]
```diff
@@ -1,2 +1,2 @@
-<a href="http://www.tx-gaming.net/warzone/tournament.php?tourney%5Bid%5D=34&amp;action=brackets" class="postlink">http://www.tx-gaming.net/warzone/tournament.php?tourney[id]=34&amp;action=brackets</a><br>
-<a href="http://www.tx-gaming.net/warzone/tournament.php?tourney%5Bid%5D=34&amp;action=brackets" class="postlink">link</a>
\ No newline at end of file
+<a href="http://example.org/?tourney%5Bid%5D=34&amp;action=brackets" class="postlink">http://example.org/?tourney[id]=34&amp;action=brackets</a><br>
```
Comment (marc1706) [inline tests/text_processing/tickets_data/PHPBB3-10587.html]: Could you add test cases with a frontend file like index.php, too?

File: phpBB/phpbb/textformatter/s9e/parser.php (hunk 3 above)
Comment (JoshyPHP) [inline phpBB/phpbb/textformatter/s9e/parser.php]: I'm sorry but this is enough of an extra burden to tip the scale in favor of abandonning this PR. I've spent several hours figuring out how the old code worked and whether it can be made testable that I'm at a point where spending another minute making comments less readable for me is excruciating. I hope you understand.

File: tests/text_processing/tickets_data/PHPBB3-10587.html (hunk 4 above)
Comment (JoshyPHP) [inline tests/text_processing/tickets_data/PHPBB3-10587.html]: This is an unrelated test whose output is affected by this patch.

**Comment by marc1706** [comment]
How are things looking for this PR? To me it seems like there is only one test that needs to be updated.

**Comment by JoshyPHP** [comment]
@marc1706 Created https://area51.phpbb.com/phpBB/viewtopic.php?f=126&t=48751 for comments

**Comment by JoshyPHP** [comment]
!unset wip

**Comment by JoshyPHP** [comment]
!unset WIP

File: tests/text_formatter/s9e/default_formatting_test.php [hunk 5]
```diff
@@ -226,6 +226,36 @@ public function get_default_formatting_tests()
...
+				'http://localhost/phpbb/viewforum.php?f=1#xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
+				'<a href="http://localhost/phpbb/viewforum.php?f=1#xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" class="postlink">viewforum.php?f=1#xxxxxxxxxxxxxxxxxxxxx ... xxxxxxxxxx</a>'
+			),
+			array(
+				'[url]http://example.org/0xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0[/url]',
+				'<a href="http://example.org/0xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0" class="postlink">http://example.org/0xxxxxxxxxxxxxxxxxxx ... xxxxxxxxx0</a>'
+			),
+			array(
+				'[URL]http://example.org/0xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0[/url]',
+				'<a href="http://example.org/0xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0" class="postlink">http://example.org/0xxxxxxxxxxxxxxxxxxx ... xxxxxxxxx0</a>'
```
Comment (Elsensee) [inline tests/text_formatter/s9e/default_formatting_test.php]: This test *should* fail now since you changed the behaviour so it doesn't shorten links in url bbcodes anymore, right?

Comment (JoshyPHP) [inline tests/text_formatter/s9e/default_formatting_test.php]: You're right. I'm surprised that the tests still passed locally.

Comment (JoshyPHP) [inline tests/text_formatter/s9e/default_formatting_test.php]: Oh, it's a funny one: the URL is shortened because URLs that appear in plain text are shortened. In hindsight, it's perfectly logical.

**Comment by JoshyPHP** [comment]
!set WIP

**Comment by JoshyPHP** [comment]
!unset WIP

**Comment by marc1706** [comment]
@JoshyPHP changed my mind: https://area51.phpbb.com/phpBB/viewtopic.php?f=126&t=48751&p=286796#p286796
Care to change it to that? :)

**Comment by JoshyPHP** [comment]
@marc1706 I removed the commit that limited shortening to linkified URLs only.
